pkg/apis/migration/migration.go

```diff
@@ -407,3 +415,129 @@ func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volu
 	}
 	return volumeID.Id, nil
 }
+
+// cleanupStaleCRDInstances helps in cleaning up stale volume migration CRD instances
+func (volumeMigration *volumeMigration) cleanupStaleCRDInstances(ctx context.Context) {
+	log := logger.GetLogger(ctx)
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		log.Errorf("failed to get kubeconfig. err: %v", err)
+		return
+	}
+	crdGroupClient, err := k8s.NewClientForGroup(ctx, config, CRDGroupName)
+	if err != nil {
+		log.Errorf("Failed to create new client for group %q . Err: %+v", CRDGroupName, err)
+		return
+	}
+	// Clean up routine to cleanup successful CnsRegisterVolume instances
+	err = volumeMigrationInstance.watcher(ctx)
+	if err != nil {
+		log.Error("Failed to watch on config file for changes to VolumeMigrationCRCleanupIntervalInHours. Error: %+v", err)
+		return
+	}
+	go func() {
+		for {
+			ctx, log = logger.GetNewContextWithLogger()
+			log.Infof("Triggering CnsVSphereVolumeMigration cleanup routine")
+			volumeMigrationResourceList := &migrationv1alpha1.CnsVSphereVolumeMigrationList{}
+			err = crdGroupClient.List(ctx, volumeMigrationResourceList)
+			if err != nil {
+				log.Warnf("failed to get CnsVSphereVolumeMigration list from the cluster. Err: %+v", err)
+			}
+			log.Debugf("CnsVSphereVolumeMigrationList: %+v", volumeMigrationResourceList)
+			queryFilter := cnstypes.CnsQueryFilter{
+				ContainerClusterIds: []string{
+					volumeMigrationInstance.cnsConfig.Global.ClusterID,
+				},
+			}
+			queryAllResult, err := (*volumeMigrationInstance.volumeManager).QueryAllVolume(ctx, queryFilter, cnstypes.CnsQuerySelection{})
+			if err != nil {
+				log.Errorf("failed to queryAllVolume with err %+v", err)
+				return
+			}
+			log.Debugf("QueryVolumeInfo successfully returned with result:  %v:", spew.Sdump(queryAllResult))
+			cnsVolumesMap := make(map[string]bool)
+			for _, vol := range queryAllResult.Volumes {
+				cnsVolumesMap[vol.VolumeId.Id] = true
+			}
+			log.Debugf("cnsVolumesMap:  %v:", cnsVolumesMap)
+			var resourcesToBeDeleted []string
+			for _, volumeMigrationResource := range volumeMigrationResourceList.Items {
+				if _, existsInCNSVolumesMap := cnsVolumesMap[volumeMigrationResource.Name]; !existsInCNSVolumesMap {
+					log.Debugf("Volume with id %s is not found in CNS", volumeMigrationResource.Name)
+					resourcesToBeDeleted = append(resourcesToBeDeleted, volumeMigrationResource.Name)
+				}
+			}
+			for _, volID := range resourcesToBeDeleted {
+				err = volumeMigrationInstance.DeleteVolumeInfo(ctx, volID)
+				if err != nil {
+					log.Warnf("failed to delete volume mapping CR for %s with error %+v", volID, err)
+					continue
+				}
+			}
+			log.Infof("Completed CnsVSphereVolumeMigration cleanup")
+			for i := 1; i <= volumeMigrationInstance.cnsConfig.Global.VolumeMigrationCRCleanupIntervalInHours; i++ {
+				time.Sleep(time.Duration(1 * time.Hour))
+			}
+		}
+	}()
+}
+
+// watcher watches on the vsphere.conf file mounted as secret
```

Comment:
As discussed we do not need `watcher` and `reloadConfiguration` here in the volume migration service. Can you remove them? We already watch for config change in metadata syncer and controller where we are initializing volume migration service.

Volume Migration Service is initialized with a pointer for volume manager and config, so we are good.

```
GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Manager, cnsConfig *cnsconfig.Config)
```

Reply:
Removed the watcher and reload config method